### DISCUSSION TITLE
Inline abstract/interface implementation

### DISCUSSION THREAD

**Discussion by tmat** [dotnet/csharplang#6049] [Language Ideas]
### Motivation

The goal is to make it possible to define an implementation of an interface or an abstract type within an expression in a succinct manner. Such feature would significantly reduce the verbosity of declaring a separate type and all the implemented method signatures. It would also enable implementing an interface or an abstract type locally, only in the scope where its needed.

Previously the concept of inner classes have been discussed but these are more verbose than the current proposal and also have unresolved issues with closures.

Generalization of https://github.com/dotnet/csharplang/discussions/2517 

### Inspiration

Currently we can write

```C#
var x = 1;
var y = 2;
var impl = new D(() => x + y);

delegate int D();
```

The compiler generates a display class (representing the closure) and `x` and `y` variables are emitted as fields of this type.
Conceptually, delegates are similar to single-method interface types and it would make sense to use similar syntax to define inline interface implementation.

### Proposal

The proposal is to use an existing _new expression_ syntax but interpret certain forms that are currently treated as semantic errors differently.

Specifically, `new T(<args>)` where `T` is an interface or an abstract type would be allowed as long as 

1) `<args>` is a single unnamed parameter and `T` has a single abstract method.
2) `<args>` is is an argument list of named arguments `name1: <expr1>, ..., nameN: <exprN>` where each `name1` is a name of an abstract method of `T` (inherited or directly declared by `T`) and `<expr1>` is a lambda declaration. If `T` only has a single overload of method `name1` then the compiler will infer the types of the lambda parameters from the signature of this overload. Otherwise, the lambda has to specify the types in its signature and the signature has to match one of the overloads. Each abstract method of `T` must match exactly one of the specified lambdas. 

If `T` inherits multiple interface methods of the same name and signature from its implemented interfaces then the lambda that matches the name of the method and its signature would implement all these methods. 

> The syntax of named argument could also be extended to allow specifying type name (e.g. `I.M`) to disambiguate between these interface methods and thus allow specifying different implementations for these methods (explicit interface method implementation).

The compiler would emit a generated private type `G` implementing `T`, if `T` is an interface, or a subtype of `T`, if `T` is an abstract type. The bodies of the specified lambdas would become bodies of methods of `G`.

Examples:

1) Single method

```C#
var x = 1;
var y = 2;
var impl = new I(() => x + y);

public interface I
{
  int GetValue();
}
```

Would be compiled into

```
[CompilerGenerated]
private sealed class G : I
{
    private __DisplayClass closure;

    public G(__DisplayClass closure)
    {
       this.closure = closure;
    }

    internal int GetValue()
    {
        return closure.x + closure.y;
    }
}

var closure = new __DisplayClass(); // standard closure instantiation
closure.x = 1;
closure.y = 2;
I impl = new G(closure);
```

2) Multiple methods

```C#
var comparer = new IEqualityComparer<C>(
  Equals: (left, right) => left.X == right.X,
  GetHashCode: obj => obj.X.GetHashCode());

class C { public int X; }
```

compiles into

```
[CompilerGenerated]
private sealed class G : IEqualityComparer<C>
{
    internal int Equals(C left, C right) { return left.X == right.X; }
    internal int GetHashCode(C obj) { return obj.X.GetHashCode(); }
}

IEqualityComparer<C> comparer = new G();
```

### Possible extensions

1) Ability to specify overrides of virtual non-abstract methods.
  If `T` is an abstract class with virtual, non-abstract method `M` the _new expression_ can specify lambda to override such method.

2) Extend named arguments syntax to allow:
  a) different implementations for inherited interface methods that have same name and signature: `MyInterface.Method: <lambda>`, `MyInterface<int>.Method: <lambda>`
  b) Writable properties: `Value.set: <lambda>`, `Value.get: <lambda>`
  c) Read-only properties can use simple name: `Value: <lambda>`,
  d) Event handlers: `E.add: <lambda>`, `E.remove: <lambda>`.
  e) Indexers: `get: <lambda>`, `set: <lambda>`

  Alternative position of `get`, `set` `add` and `remove` keywords:
    `get Value: <lambda>`, `set Value: <lambda>`, `add E: <lamda>`, `remove E: <lambda>`

3) If `T` is an interface the generated type can be a struct. In constrained context this would allow avoiding allocation: 
See https://github.com/dotnet/csharplang/discussions/6049#discussioncomment-2612133

**Comment by HaloFour** [comment]
#2517  😁

**Reply by HaloFour** [reply]
Basically the emitted lambda displayclass would simply implement the target interface, correct?  ~Would this apply to any SAM type (single abstract method) or only interfaces?~ (nevermind, I should probably read the title!)

The syntax for the multiple method case feels a little weird to me, specifically that it uses constructor + named argument syntax.  The use of named arguments makes it feel like one or more of the members should be optional, or that the names could be omitted for some form of positional construction, but I assume neither would be the case.  I think I'd prefer a nominal form, e.g.:

```cs
// like methods
var comparer = new IEqualityComparer<C> {
    Equals(left, right) => left.X == right.X,
    GetHashCode(obj) => obj.X.GetHashCode();
};

// or like object initializers with lambdas
var comparer = new IEqualityComparer<C> {
    Equals = (left, right) => left.X == right.X,
    GetHashCode = (obj) => obj.X.GetHashCode();
};

// or like object initializers with event handlers
var comparer = new IEqualityComparer<C> {
    Equals += (left, right) => left.X == right.X,
    GetHashCode += (obj) => obj.X.GetHashCode();
};
```

Whatever the syntax is I think it would have to work well with abstract types that could have either/both constructors with required parameters and object initializers:

```cs
abstract class Foo {
    public abstract int Add(int x, int y);
    int Value { get; set; }

    protected Foo(int x, int y) { }
}

// spaghetti
var foo = new Foo(2, 3) {
    Value = 4,
    Add += (x, y) => x + y
};
```

But either way, I'm excited that the concept is gaining traction.

**Reply by tmat** [reply]
@HaloFour Thanks for the pointer to #2517. Seems like we can treat this proposal as an extension of yours :).

I chose the named arguments since this is already a valid syntax. You can implement abstract types that make it easy to use delegates to implement the abstract methods, it'd just be better if the boilerplate was emitted by the compiler and without the overhead of delegate allocations and extra delegate dispatch.

```C#
class DelegatingEqualityComparer<T> : IEqualityComparer<T>
{
  Func<T, T, bool> _equals;
  Func<T, int> _getHashCode;

  DelegatingEqualityComparer(Func<T, T, bool> Equals, Func<T, int> GetHashCode)
  {
    _equals = Equals;
    _getHashCode = GetHashCode;
  }

  public bool Equals(T left, T right) => _equals(left, right);
  public bool GetHashCode(T obj) => _getHashCode(obj);
}

var comparer = new DelegatingEqualityComparer<C>(
  Equals: (left, right) => left.X == right.X,
  GetHashCode: obj => obj.X.GetHashCode());
```

**Reply by tmat** [reply]
We can also take another step and allow to specify overrides of virtual non-abstract methods. Those would be truly optional.

**Reply by tmat** [reply]
A couple of things that we would need an extended syntax over what's currently allowed by named arguments syntax are:
1) different implementations for inherited interface methods that have same name and signature: `Iface.Method: <lambda>`
2) property setters: perhaps `Value.set: <lambda>`, `Value.get: <lambda>`, event handlers `E.add: <lambda>`, `E.remove: <lambda>`.

**Reply by jnm2** [reply]
What about syntax like this?

```cs
var x = new struct ISomething
{
    M(int p) => ...;
    M(string p) => ...;
    SomeProperty { get => ...; set => ...; }
    SomeEvent { add => ...; remove => ...; }
    this[int index] { get => ...; set => ...; }
};

**Reply by tmat** [reply]
Definitely possible, just needs more specialized and also verbose syntax.
I think we should optimize the syntax for methods and read-only properties.

**Reply by HaloFour** [reply]
@tmat 

> I chose the named arguments since this is already a valid syntax.

I think that's actually a liability as it immediately complicates how you could pass arguments to a base constructor for an abstract class.  You might have a mix of positional and named arguments as well as these named methods, and the latter two could theoretically have the same name:

```cs
abstract class Foo {
    protected Foo(string X, int Y) { ... }
    public abstract void X();
}

var foo = new Foo(Y: 1, X: "value", X: () => { }); // uh wha?
```

Moving them to object initializer syntax makes that a lot less confusing to me, and would also eliminate the possibility of name conflicts.

**Comment by tannergooding** [comment]
I would like to see a syntax that optionally allows the backing type to be a struct, not a class. The pattern of `where T : ISomeInterface` allows for generic specialization when `T` is a struct and therefore for the JIT to do more optimization and even inlining, which is particularly useful for things like small comparers, predicates, etc.

Having this be two allocations (the closure and then the implementing type) makes it significantly less practical for many types of code.

It would be nice to see some syntax that allows you to avoid capturing (e.g. `static`) where relevant.

**Reply by tmat** [reply]
> I would like to see a syntax that optionally allows the backing type to be a struct, not a class

`var x = new struct I(AddOne: a => a+1)`?

> Having this be two allocations (the closure and then the implementing type) makes it significantly less practical for many types of code.

Perhaps the compiler might be able to make the closure implement the interface as an optimization.

**Reply by tmat** [reply]
> It would be nice to see some syntax that allows you to avoid capturing (e.g. static) where relevant.

The lambdas can be static using current syntax if you want to explicitly prevent closures:

`var x = new I(m: static a => a+1)`

In any case, if no variable is lifted there would be no closure.

**Reply by tannergooding** [reply]
Yup, just wanted to call out the desire for this to be extensible to one of the common "perf oriented" patterns.

The pattern of `public static int BinarySearch<T, TComparable>(
this Span<T> span, TComparable comparable)
 where TComparable : IComparable<T>` isn't uncommon (we expose it in several places in the BCL) and it allows the best code-generation (particularly when `TComparer` is a `struct`).

The major downside is that its not nearly as trivial as just `span.BinarySearch((x, y) => x < y)` or `span.BinarySearch(Comparer<T>.Default)` and users have to go implement an entire type just to get this efficiency. Having some shorthand like `span.BinarySearch(new struct IComparer<T>((x, y) => x < y))` might be reasonable enough for this to become "more common place" and there-by allow better perf with roughly the same "easy" syntax users already like.

**Reply by CyrusNajmabadi** [reply]
Note: `struct IComparer<T>` also matches something i want with `yield` enumerators.  Namely at a signature level i would like to write:  `public struct IEnumerable<T> GetValues() { ... yield ... }` indicating that i'm getting some anonymous impl back, but that you can know that that anonymous impl is a struct and does implement that shape.

**Reply by tmat** [reply]
Couldn't the compiler always emit struct when implementing an interface (without explicit syntax)? If it is not constrained it will get boxed, but that's the same as class then.

**Reply by HaloFour** [reply]
```cs
var x = new static struct I(AddOne: a => a+1);
```

**Reply by jnm2** [reply]
What about anonymous interfaces as type parameters:

```cs
DoSomeCounting<ISomeKindOfNumber { static Zero: 0, static AddOne: a => a + 1 }>();
```

**Reply by CyrusNajmabadi** [reply]
> Couldn't the compiler always emit struct when implementing an interface (without explicit syntax)? If it is not constrained it will get boxed, but that's the same as class then.

In this case, i mean that we woudl not box.  The "IEnumerable" part would just be stating that the struct implements that interface, but the signature would refer to the unnuterable struct name.  But i'm afraid i'm digressing this far off the original intent of this discussion :)

**Reply by tmat** [reply]
@CyrusNajmabadi Yup, my comment is meant to apply to this proposal only.

**Comment by En3Tho** [comment]
I think you should also include:
1. Multiple interfaces case, like new I1, I2.
2. A case when overloading fails (interface has overloaded methods)

**Comment by HaloFour** [comment]
To be honest I think you'd cover the majority of the use cases by having the compiler support lambda syntax for SAM types (single abstract method).  You wouldn't have to worry about any new syntax, and IMO the code generation is also pretty simple.  At most I think the concerns would be around method overloading as clearly a lambda would have to favor a delegate overload to retain backward compatibility.

I'm not arguing against multiple methods or more full-fledged anonymous implementation, only that this part of the proposal is more complicated and carries more risk, thus it might be worthwhile to consider the two as separate.

**Reply by HaloFour** [reply]
I also wonder if lambda support for SAM types could be the basis for value delegates.  Given a method signature like:

```cs
public C<TR> Map<TR, F>(F fn) where F : struct, IFunc<T, TR> { ... }
```

the compiler could emit a struct version of the display type implementing that interface and pass that as `F`:

```cs
C<int> c1 = ...;
C<string> c2 = c1.Map(x => x.ToString());
```

which would be compiled into:

```cs
struct LambdaDisplayStruct : IFunc<int, string> {
    public string Apply(int arg) => arg.ToString();
}

var c2 = c1.Map<int, LambdaDisplayStruct>(default);
```

**Comment by alrz** [comment]
I think I'd rather have local class declarations to avoid all the new syntax plus inline declarations tend to get hard to read, IMO.

```cs
I M() {
  class C : I {} // zero syntax addition
  return new C();
}
```
Does the anonymity really matter here?

**Reply by HaloFour** [reply]
IMO local types don't buy you much.  They're just as verbose, and only slightly more local.  I also disagree that they're harder to read, but that may depend on the syntax chosen here.  Lambdas have already demonstrated that an inline form is easier to use.  For any language I've worked with that offers both forms, local types are _very_ rarely used compared to anonymous implementations.

**Reply by Richiban** [reply]
Also, the syntax for local types clashes with that of nested types (in the case of a class container), which might cause problems?

**Comment by Spacefish** [comment]
I guess the most prominent use case for a feature like that would be passing a callback handler to a function.
So imagine something like this:
```csharp
public interface IMyFunctionCallback
{
  void CaseA(int data);
  void CaseB();
  void OnError(string error);
}

public void MyFunction(IMyFunctionCallback callback) { .... }
```

you could call MyFunction + supply multiple typed handlers via
```csharp
MyFunction(new {
   void CaseA(int data) { Console.WriteLine($"case A data is: {data}"); }
   void CaseB() { Console.WriteLine("case B"); }
   void OnError(string msg) { Console.WriteLine($"An error occured: {msg}"); }
} : IMyFunctionCallback);
```

just like lambdas but having them not as different parameters of the funciton but described in an interface.
Could be usefull if you have multiple function which all need the same callback handlers, as you don´t have to repeat the callback function parameters on all functions + as a caller you could have a generic callback implementation which can be passed arround as just one reference.

i think the syntax

```csharp
new { ... } : IntefaceA, InterfaceB
```
is the one i prefer the most personally..

I think
```csharp
new  : InterfaceA, InterfaceB { ... }
```
looks akward due to the `new` followed by a `:`

B.t.w. why not even allow anonymous classes to extend existing ones like:

```csharp
public class MyBaseClass {
  public int Value => 123;
}

new {
  int SquaredValue => Value*Value;
} : MyBaseClass
```

**Reply by HaloFour** [reply]
IMO those cases are better solved by having an overload of `MyFunction` that accepts three separate callback delegates, then you can call it using existing lambda syntax:

```cs
MyFunction(
    data => Console.WriteLine($"case A data is: {data}"),
    () => Console.WriteLine("case B"),
    msg => Console.WriteLine($"An error occured: {msg}")
);
```

This has become a more common pattern in Java as callbacks have migrated more to functional interfaces which can be implemented via lambdas.

**Reply by Spacefish** [reply]
Yeah that´s definetly possible, but a drawback would be if you have multiple functions with the same callback types, you would have to write:

```csharp
MyFunctionA(CallbackA a, CallbackB b, CallbackC c) { ... }
MyFunctionB(CallbackA a, CallbackB b, CallbackC c) { ... }
MyFunctionC(CallbackA a, CallbackB b, CallbackC c) { ... }
```

instead of
```csharp
MyFunctionA(ISpecialCallback callback) { ... }
MyFunctionB(ISpecialCallback callback) { ... }
MyFunctionC(ISpecialCallback callback) { ... }
```

same thing when calling them.. In the later case you have to pass one argument to a type implementing ISpecialCallback in the former case 3 references to your three different callbacks.

**Reply by HaloFour** [reply]
You could have a helper method that creates an instance of `ISpecialCallback` that wraps the three lambda instances:

```cs
ISpecialCallback callback = SpecialCallback.Of(
    a: data => Console.WriteLine($"case A data is: {data}"),
    b: () => Console.WriteLine("case B"),
    c: msg => Console.WriteLine($"An error occured: {msg}")
);

MyFunctionA(callback);
```

I personally don't find having to accept/pass multiple arguments as much of a drawback.  As mentioned, this pattern has become much more common in Java and has largely supplanted callback interfaces that have multiple methods for these scenarios.

**Comment by hovik-aghajanyan** [comment]
There are many use cases for this, and inline instance creation will reduce multiple dummy class creation in many simple cases like this.

```
interface IDateProvider
{
      DateTime NextDate {get;}
}
```

instead of creating a new class, we can use like this

```
service.AddScoped<IDateProvider>(new IDateProvider
{
   public DateTime NextDate => DateTime.Now.AddDay(1);
}
```
which can be replaced at compile time with this piece of code

```
file class DataProvider+SomeGuid: IDateProvider {
     public DateTime NextDate => DateTime.Now.AddDay(1);
}

service.AddScoped<IDateProvider>(new  DataProvider+SomeGuid())
```
